File: node/ts/zkgroup/groupsend/GroupSendCredential.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { randomBytes } from 'crypto';

import ByteArray from '../internal/ByteArray';
import { RANDOM_LENGTH } from '../internal/Constants';
import * as Native from '../../../Native';

import GroupSendCredentialPresentation from './GroupSendCredentialPresentation';
import type GroupSendCredentialResponse from './GroupSendCredentialResponse'; // for docs
import ServerPublicParams from '../ServerPublicParams';

/**
 * A credential indicating membership in a group, based on the set of *other* users in the
 * group with you.
 *
 * Follows the usual zkgroup pattern of "issue response -> receive response -> present credential
 * -> verify presentation".
 *
 * @see {@link GroupSendCredentialResponse}
 * @see {@link GroupSendCredentialPresentation}
 */
export default class GroupSendCredential extends ByteArray {
  private readonly __type?: never;

  constructor(contents: Buffer) {
    super(contents, Native.GroupSendCredential_CheckValidContents);
  }

  /**
   * Generates a new presentation, so that multiple uses of this credential are harder to link.
   */
  present(serverParams: ServerPublicParams): GroupSendCredentialPresentation {
    const random = randomBytes(RANDOM_LENGTH);
    return this.presentWithRandom(serverParams, random);
  }

  /**
   * Generates a new presentation with a dedicated source of randomness.
   *
   * Should only be used for testing purposes.
   *
   * @see {@link GroupSendCredential#present}
   */
  presentWithRandom(
    serverParams: ServerPublicParams,
    random: Buffer
  ): GroupSendCredentialPresentation {
    return new GroupSendCredentialPresentation(
      Native.GroupSendCredential_PresentDeterministic(
        this.contents,
        serverParams.contents,
        random
      )
    );
  }
}

File: java/shared/java/org/signal/libsignal/zkgroup/groupsend/GroupSendCredential.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.groupsend;

import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.ServerPublicParams;
import org.signal.libsignal.zkgroup.internal.ByteArray;

/**
 * A credential indicating membership in a group, based on the set of <em>other</em> users in the
 * group with you.
 *
 * <p>Follows the usual zkgroup pattern of "issue response -> receive response -> present credential
 * -> verify presentation".
 *
 * @see GroupSendCredentialResponse
 * @see GroupSendCredentialPresentation
 */
public final class GroupSendCredential extends ByteArray {

  public GroupSendCredential(byte[] contents) throws InvalidInputException {
    super(contents);
    Native.GroupSendCredential_CheckValidContents(contents);
  }

  /** Generates a new presentation, so that multiple uses of this credential are harder to link. */
  public GroupSendCredentialPresentation present(ServerPublicParams serverParams) {
    return present(serverParams, new SecureRandom());
  }

  /**
   * Generates a new presentation with a dedicated source of randomness.
   *
   * <p>Should only be used for testing purposes.
   *
   * @see #present(ServerPublicParams)
   */
  public GroupSendCredentialPresentation present(
      ServerPublicParams serverParams, SecureRandom secureRandom) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        Native.GroupSendCredential_PresentDeterministic(
            getInternalContentsForJNI(), serverParams.getInternalContentsForJNI(), random);

    try {
      return new GroupSendCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}
